v0/destinations/personalize/transform.js

```diff
@@ -29,62 +34,57 @@ async function process(event) {
         !(
           keyField[k].toUpperCase() === "USER_ID" ||
           keyField[k].toUpperCase() === "EVENT_TYPE" ||
-          keyField[k].toUpperCase() === "TIMESTAMP"
+          keyField[k].toUpperCase() === "TIMESTAMP" ||
+          keyField[k].toUpperCase() === "ITEM_ID" ||
+          keyField[k].toUpperCase() === "EVENT_VALUE" ||
+          keyField[k].toUpperCase() === "IMPRESSION" ||
+          keyField[k].toUpperCase() === "RECOMMENDATION_ID" ||
+          keyField[k].toUpperCase() === "EVENT_ID"
         )
       ) {
         const mappedFields = mappedField[k];
-
         keyField[k] = keyField[k]
           .toLowerCase()
-          .replace(/_([a-z])/g, function (g) {
+          .replace(/_([a-z])/g, function camelCase(g) {
             return g[1].toUpperCase();
           });
-        if (message.properties && message.properties[mappedFields]) {
-          property[keyField[k]] = message.properties[mappedFields];
+        if (properties && properties[mappedFields]) {
+          property[keyField[k]] = properties[mappedFields];
         } else if (typeof message[mappedFields] !== "undefined") {
           property[keyField[k]] = message[mappedFields];
         } else {
           throw new Error(`Mapped Field ${mappedFields} not found`);
         }
+      } else if (
+        !(
+          keyField[k].toUpperCase() === "USER_ID" ||
+          keyField[k].toUpperCase() === "EVENT_TYPE"
+        )
+      ) {
+        keyField[k] = keyField[k]
+          .toLowerCase()
+          .replace(/_([a-z])/g, function camelCase(g) {
+            return g[1].toUpperCase();
+          });
+        const mappedFields = mappedField[k];
+        eventObj[keyField[k]] = properties[mappedFields];
       }
     }
-    const eventList = [];
-    const eventObj = {};
-
-    if (property.eventValue) {
-      eventObj.eventValue = property.eventValue;
-      delete property.eventValue;
-    }
-    if (property.impression) {
-      eventObj.impression = property.impression;
-      delete property.impression;
-    }
-    if (property.itemId) {
-      eventObj.itemId = property.itemId;
-      delete property.itemId;
-    }
-    if (property.recommendationId) {
-      eventObj.recommendationId = property.recommendationId;
-      delete property.recommendationId;
-    }
-    let eventId;
-    if (property.eventId) {
-      eventId = property.eventId;
-      delete property.eventId;
-    } else {
-      eventId = event.message.messageId;
+    if (!property.itemId) {
```

Comment:
according to your logic now itemId will never be present then why this if?

Reply:
we cannot send data to personalize without itemId. So even if the user does not set any data in itemId, through the properties section, we have to pass some kind of data there